File: packages/crypto/src/ratchet.ts
import { fromBase64, toBase64 } from "@screenmesh/protocol";

/**
 * A simplified Double Ratchet (Signal-style) session between exactly two
 * devices, giving each envelope its own single-use encryption key —
 * genuine per-message forward secrecy, not just per-workspace-epoch.
 * See docs/Security.md §5 for the full design rationale.
 *
 * Bootstrap ("X3DH-lite"): the initial root key mixes each device's
 * long-term X25519 identity key (learned via the relay's HTTP join API —
 * NOT independently authenticated) with the pairing secret that traveled
 * over the out-of-band QR channel:
 *
 *   rootKey0 = HKDF(salt = pairingSecret, ikm = ECDH(myIdentity, theirIdentity), info = ...)
 *
 * A relay that substitutes identity keys in transit still can't derive
 * rootKey0 without the QR secret, so the QR remains the real trust
 * anchor. Both sides seed their ratchet keypair with their OWN identity
 * keypair (not a fresh ephemeral one) so EITHER side can send first
 * without having received anything — the first message(s) on a session,
 * before the peer has ever replied, therefore derive their key from
 * long-term identity material and don't have independent forward secrecy
 * from a future identity-key compromise. The moment the peer replies
 * (any message at all), both sides roll over to fresh ephemeral keys and
 * full forward secrecy applies from then on — the session "heals" after
 * one round trip. This is a deliberate, documented simplification, not
 * an oversight: a full X3DH needs a published one-time-prekey bundle,
 * which needs server infrastructure we don't have yet.
 *
 * Also deliberately out of scope: skipped-message-key tolerance is
 * capped rather than unbounded (see MAX_SKIPPED_KEYS/MAX_SKIP_PER_STEP).
 */

const HKDF_INFO_ROOT = "screenmesh-ratchet-root-v1";
const HKDF_INFO_DH = "screenmesh-ratchet-dh-v1";
const MAX_SKIPPED_KEYS = 50;
const MAX_SKIP_PER_STEP = 500;

export interface RatchetKeyPair {
  publicKey: CryptoKey;
  privateKey: CryptoKey;
}

interface SkippedKey {
  ratchetPublicKeyB64: string;
  messageNumber: number;
  messageKeyB64: string;
}

export interface RatchetSession {
  rootKey: Uint8Array;
  /** My current DH ratchet keypair — starts as my identity keypair, then
   *  becomes fresh ephemeral material the first time I have to respond
   *  to the peer's ratchet key changing. */
  myRatchetKeyPair: RatchetKeyPair;
  /** The peer's most recently observed ratchet public key. */
  theirRatchetPublicKeyB64: string;
  sendChainKey: Uint8Array | null;
  recvChainKey: Uint8Array | null;
  sendMessageNumber: number;
  recvMessageNumber: number;
  previousSendChainLength: number;
  skippedKeys: SkippedKey[];
}

export interface RatchetMessageHeader {
  ratchetPublicKeyB64: string;
  messageNumber: number;
  previousChainLength: number;
}

export class RatchetError extends Error {}

// --- primitives ---

async function ecdhBits(
  myPrivate: CryptoKey,
  theirPublic: CryptoKey,
  bitLength = 256,
): Promise<Uint8Array> {
  const bits = await crypto.subtle.deriveBits(
    { name: "X25519", public: theirPublic },
    myPrivate,
    bitLength,
  );
  return new Uint8Array(bits);
}

async function hkdf(
  ikm: Uint8Array,
  salt: Uint8Array,
  info: string,
  lengthBytes: number,
): Promise<Uint8Array> {
  const key = await crypto.subtle.importKey("raw", ikm as BufferSource, "HKDF", false, [
    "deriveBits",
  ]);
  const bits = await crypto.subtle.deriveBits(
    {
      name: "HKDF",
      hash: "SHA-256",
      salt: salt as BufferSource,
      info: new TextEncoder().encode(info),
    },
    key,
    lengthBytes * 8,
  );
  return new Uint8Array(bits);
}

async function hmacSha256(keyBytes: Uint8Array, data: Uint8Array): Promise<Uint8Array> {
  const key = await crypto.subtle.importKey(
    "raw",
    keyBytes as BufferSource,
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );
  const sig = await crypto.subtle.sign("HMAC", key, data as BufferSource);
  return new Uint8Array(sig);
}

/** One step of the symmetric-key ratchet: derive a message key, advance the chain. */
async function stepChain(
  chainKey: Uint8Array,
): Promise<{ messageKey: Uint8Array; nextChainKey: Uint8Array }> {
  const [messageKey, nextChainKey] = await Promise.all([
    hmacSha256(chainKey, new Uint8Array([0x01])),
    hmacSha256(chainKey, new Uint8Array([0x02])),
  ]);
  return { messageKey, nextChainKey };
}

async function importAesKey(bytes: Uint8Array): Promise<CryptoKey> {
  return crypto.subtle.importKey("raw", bytes as BufferSource, "AES-GCM", false, [
    "encrypt",
    "decrypt",
  ]);
}

export async function generateRatchetKeyPair(): Promise<RatchetKeyPair> {
  const pair = (await crypto.subtle.generateKey("X25519", false, ["deriveBits"])) as CryptoKeyPair;
  return { publicKey: pair.publicKey, privateKey: pair.privateKey };
}

async function exportRatchetPublicKey(key: CryptoKey): Promise<string> {
  return toBase64(new Uint8Array(await crypto.subtle.exportKey("raw", key)));
}

async function importRatchetPublicKey(b64: string): Promise<CryptoKey> {
  return crypto.subtle.importKey("raw", fromBase64(b64) as BufferSource, "X25519", true, []);
}

/**
 * Bootstrap a brand-new session with a peer. Both sides compute this with
 * the same sorted inputs and arrive at the same root key. Each side's
 * `myRatchetKeyPair` starts as ITS OWN identity keypair (not a fresh
 * ephemeral one) specifically so either side can send first without
 * having received anything yet — see the module doc comment for the
 * forward-secrecy trade-off this implies until the first round trip.
 */
export async function initRatchetSession(params: {
  workspaceId: string;
  myDeviceId: string;
  myIdentityPublic: CryptoKey;
  myIdentityPrivate: CryptoKey;
  peerDeviceId: string;
  peerIdentityPublic: CryptoKey;
  pairingSecret: Uint8Array;
}): Promise<RatchetSession> {
  const ecdh = await ecdhBits(params.myIdentityPrivate, params.peerIdentityPublic);
  const [a, b] = [params.myDeviceId, params.peerDeviceId].sort();
  const info = `${HKDF_INFO_ROOT}|${params.workspaceId}|${a}|${b}`;
  const rootKey = await hkdf(ecdh, params.pairingSecret, info, 32);
  return {
    rootKey,
    myRatchetKeyPair: { publicKey: params.myIdentityPublic, privateKey: params.myIdentityPrivate },
    theirRatchetPublicKeyB64: await exportRatchetPublicKey(params.peerIdentityPublic),
    sendChainKey: null,
    recvChainKey: null,
    sendMessageNumber: 0,
    recvMessageNumber: 0,
    previousSendChainLength: 0,
    skippedKeys: [],
  };
}

/** Advance the DH ratchet: mix a fresh DH result into the root key, derive a new chain key. */
async function dhRatchetStep(
  rootKey: Uint8Array,
  myPrivate: CryptoKey,
  theirPublic: CryptoKey,
): Promise<{ rootKey: Uint8Array; chainKey: Uint8Array }> {
  const dh = await ecdhBits(myPrivate, theirPublic);
  const out = await hkdf(dh, rootKey, HKDF_INFO_DH, 64);
  return { rootKey: out.slice(0, 32), chainKey: out.slice(32, 64) };
}

function pushSkipped(session: RatchetSession, entry: SkippedKey): void {
  session.skippedKeys.push(entry);
  while (session.skippedKeys.length > MAX_SKIPPED_KEYS) session.skippedKeys.shift();
}

function takeSkipped(
  session: RatchetSession,
  ratchetPublicKeyB64: string,
  messageNumber: number,
): Uint8Array | null {
  const idx = session.skippedKeys.findIndex(
    (k) => k.ratchetPublicKeyB64 === ratchetPublicKeyB64 && k.messageNumber === messageNumber,
  );
  if (idx === -1) return null;
  const found = session.skippedKeys[idx];
  session.skippedKeys.splice(idx, 1);
  return fromBase64(found!.messageKeyB64);
}

/** Advance recvChainKey forward to `upTo`, caching each derived key as skipped (not returned). */
async function skipForward(
  session: RatchetSession,
  ratchetPublicKeyB64: string,
  upTo: number,
): Promise<void> {
  if (!session.recvChainKey) return;
  const count = upTo - session.recvMessageNumber;
  if (count <= 0) return;
  if (count > MAX_SKIP_PER_STEP) {
    throw new RatchetError(`refusing to derive ${count} skipped keys in one step`);
  }
  for (let i = 0; i < count; i++) {
    const { messageKey, nextChainKey } = await stepChain(session.recvChainKey);
    pushSkipped(session, {
      ratchetPublicKeyB64,
      messageNumber: session.recvMessageNumber + i,
      messageKeyB64: toBase64(messageKey),
    });
    session.recvChainKey = nextChainKey;
  }
  session.recvMessageNumber = upTo;
}

/**
 * Sender side: derive the next message key, mutating the session (the
 * chain advances one-way; the returned key is meant to be used once).
 * Callers MUST persist the mutated session after this returns.
 */
export async function ratchetEncrypt(
  session: RatchetSession,
): Promise<{ messageKey: CryptoKey; header: RatchetMessageHeader }> {
  if (!session.sendChainKey) {
    const theirPublic = await importRatchetPublicKey(session.theirRatchetPublicKeyB64);
    const stepped = await dhRatchetStep(
      session.rootKey,
      session.myRatchetKeyPair.privateKey,
      theirPublic,
    );
    session.rootKey = stepped.rootKey;
    session.sendChainKey = stepped.chainKey;
    session.previousSendChainLength = session.sendMessageNumber;
    session.sendMessageNumber = 0;
  }
  const { messageKey, nextChainKey } = await stepChain(session.sendChainKey);
  const header: RatchetMessageHeader = {
    ratchetPublicKeyB64: await exportRatchetPublicKey(session.myRatchetKeyPair.publicKey),
    messageNumber: session.sendMessageNumber,
    previousChainLength: session.previousSendChainLength,
  };
  session.sendChainKey = nextChainKey;
  session.sendMessageNumber += 1;
  return { messageKey: await importAesKey(messageKey), header };
}

/**
 * Receiver side: derive the message key matching this header, mutating
 * the session (performing a DH ratchet step if the peer's key changed —
 * or if we've never received anything yet — and caching any skipped-over
 * keys for later out-of-order arrivals). Callers MUST persist the
 * mutated session after this returns.
 */
export async function ratchetDecrypt(
  session: RatchetSession,
  header: RatchetMessageHeader,
): Promise<CryptoKey> {
  const cached = takeSkipped(session, header.ratchetPublicKeyB64, header.messageNumber);
  if (cached) return importAesKey(cached);

  if (header.ratchetPublicKeyB64 !== session.theirRatchetPublicKeyB64 || !session.recvChainKey) {
    if (session.recvChainKey) {
      // Drain whatever's left of the OLD receiving chain (bounded by the
      // sender's declared previous-chain length) so late arrivals on it
      // can still be found in the skipped-key cache.
      await skipForward(session, session.theirRatchetPublicKeyB64, header.previousChainLength);
    }
    const theirNewPublic = await importRatchetPublicKey(header.ratchetPublicKeyB64);
    const stepped = await dhRatchetStep(
      session.rootKey,
      session.myRatchetKeyPair.privateKey,
      theirNewPublic,
    );
    session.rootKey = stepped.rootKey;
    session.recvChainKey = stepped.chainKey;
    session.recvMessageNumber = 0;
    session.theirRatchetPublicKeyB64 = header.ratchetPublicKeyB64;
    // Force a fresh sending step next time we send, with fresh ephemeral
    // material — this is what "heals" the session after a round trip.
    session.sendChainKey = null;
    session.myRatchetKeyPair = await generateRatchetKeyPair();
  }

  if (!session.recvChainKey) {
    throw new RatchetError("no receiving chain established for this header");
  }
  if (header.messageNumber < session.recvMessageNumber) {
    throw new RatchetError(
      `message ${header.messageNumber} on the current chain was already consumed and is not in the skipped-key cache`,
    );
  }
  if (header.messageNumber > session.recvMessageNumber) {
    await skipForward(session, header.ratchetPublicKeyB64, header.messageNumber);
  }
  const { messageKey, nextChainKey } = await stepChain(session.recvChainKey);
  session.recvChainKey = nextChainKey;
  session.recvMessageNumber = header.messageNumber + 1;
  return importAesKey(messageKey);
}

File: apps/android/app/src/main/java/com/screenmesh/crypto/Ratchet.kt
package com.screenmesh.crypto

import org.bouncycastle.crypto.agreement.X25519Agreement
import org.bouncycastle.crypto.digests.SHA256Digest
import org.bouncycastle.crypto.generators.HKDFBytesGenerator
import org.bouncycastle.crypto.generators.X25519KeyPairGenerator
import org.bouncycastle.crypto.params.HKDFParameters
import org.bouncycastle.crypto.params.X25519KeyGenerationParameters
import org.bouncycastle.crypto.params.X25519PrivateKeyParameters
import org.bouncycastle.crypto.params.X25519PublicKeyParameters
import java.security.SecureRandom
import javax.crypto.Mac
import javax.crypto.SecretKey
import javax.crypto.spec.SecretKeySpec

/**
 * Kotlin mirror of packages/crypto/src/ratchet.ts — a simplified Double
 * Ratchet (Signal-style) session between exactly two devices, giving each
 * envelope its own single-use encryption key. See the TS file's module
 * doc comment (and docs/Security.md §5) for the full design rationale,
 * including the deliberate X3DH-lite bootstrap — both sides seed
 * `myRatchetKeyPair` with their OWN identity keypair so either can send
 * first, the session "heals" to fresh ephemeral keys after one round
 * trip — and the capped skipped-message-key tolerance
 * (MAX_SKIPPED_KEYS/MAX_SKIP_PER_STEP).
 *
 * RatchetSession is a mutable class, not an immutable data class the TS
 * side reassigns fields on: since it's a plain reference type, mutating
 * it in place IS persisting it for any caller holding the same reference
 * — there's no separate "save the returned session" step needed here.
 */

private const val HKDF_INFO_ROOT = "screenmesh-ratchet-root-v1"
private const val HKDF_INFO_DH = "screenmesh-ratchet-dh-v1"
private const val MAX_SKIPPED_KEYS = 50
private const val MAX_SKIP_PER_STEP = 500

class RatchetError(message: String) : Exception(message)

data class RatchetKeyPair(
    val publicKey: X25519PublicKeyParameters,
    val privateKey: X25519PrivateKeyParameters,
)

private data class SkippedKey(
    val ratchetPublicKeyB64: String,
    val messageNumber: Int,
    val messageKeyB64: String,
)

data class RatchetMessageHeader(
    val ratchetPublicKeyB64: String,
    val messageNumber: Int,
    val previousChainLength: Int,
)

class RatchetSession(
    var rootKey: ByteArray,
    /** My current DH ratchet keypair — starts as my identity keypair, then
     *  becomes fresh ephemeral material the first time I have to respond
     *  to the peer's ratchet key changing. */
    var myRatchetKeyPair: RatchetKeyPair,
    /** The peer's most recently observed ratchet public key. */
    var theirRatchetPublicKeyB64: String,
    var sendChainKey: ByteArray?,
    var recvChainKey: ByteArray?,
    var sendMessageNumber: Int,
    var recvMessageNumber: Int,
    var previousSendChainLength: Int,
) {
    internal val skippedKeys: MutableList<SkippedKey> = mutableListOf()
}

// --- primitives ---

private fun ecdhBits(myPrivate: X25519PrivateKeyParameters, theirPublic: X25519PublicKeyParameters): ByteArray {
    val agreement = X25519Agreement()
    agreement.init(myPrivate)
    val out = ByteArray(agreement.agreementSize)
    agreement.calculateAgreement(theirPublic, out, 0)
    return out
}

private fun hkdf(ikm: ByteArray, salt: ByteArray, info: String, lengthBytes: Int): ByteArray {
    val generator = HKDFBytesGenerator(SHA256Digest())
    generator.init(HKDFParameters(ikm, salt, info.toByteArray(Charsets.UTF_8)))
    val out = ByteArray(lengthBytes)
    generator.generateBytes(out, 0, lengthBytes)
    return out
}

private fun hmacSha256(keyBytes: ByteArray, data: ByteArray): ByteArray {
    val mac = Mac.getInstance("HmacSHA256")
    mac.init(SecretKeySpec(keyBytes, "HmacSHA256"))
    return mac.doFinal(data)
}

/** One step of the symmetric-key ratchet: derive a message key, advance the chain. */
private fun stepChain(chainKey: ByteArray): Pair<ByteArray, ByteArray> {
    val messageKey = hmacSha256(chainKey, byteArrayOf(0x01))
    val nextChainKey = hmacSha256(chainKey, byteArrayOf(0x02))
    return messageKey to nextChainKey
}

private fun importAesKey(bytes: ByteArray): SecretKey = SecretKeySpec(bytes, "AES")

fun generateRatchetKeyPair(): RatchetKeyPair {
    val gen = X25519KeyPairGenerator()
    gen.init(X25519KeyGenerationParameters(SecureRandom()))
    val pair = gen.generateKeyPair()
    return RatchetKeyPair(
        publicKey = pair.public as X25519PublicKeyParameters,
        privateKey = pair.private as X25519PrivateKeyParameters,
    )
}

private fun exportRatchetPublicKey(key: X25519PublicKeyParameters): String = toBase64(key.encoded)

private fun importRatchetPublicKey(b64: String): X25519PublicKeyParameters =
    X25519PublicKeyParameters(fromBase64(b64), 0)

/**
 * Bootstrap a brand-new session with a peer. Both sides compute this with
 * the same sorted inputs and arrive at the same root key. `myRatchetKeyPair`
 * starts as MY OWN identity keypair (not a fresh ephemeral one) so either
 * side can send first without having received anything yet.
 */
fun initRatchetSession(
    workspaceId: String,
    myDeviceId: String,
    myIdentityPublic: X25519PublicKeyParameters,
    myIdentityPrivate: X25519PrivateKeyParameters,
    peerDeviceId: String,
    peerIdentityPublic: X25519PublicKeyParameters,
    pairingSecret: ByteArray,
): RatchetSession {
    val ecdh = ecdhBits(myIdentityPrivate, peerIdentityPublic)
    val sorted = listOf(myDeviceId, peerDeviceId).sorted()
    val info = "$HKDF_INFO_ROOT|$workspaceId|${sorted[0]}|${sorted[1]}"
    val rootKey = hkdf(ecdh, pairingSecret, info, 32)
    return RatchetSession(
        rootKey = rootKey,
        myRatchetKeyPair = RatchetKeyPair(myIdentityPublic, myIdentityPrivate),
        theirRatchetPublicKeyB64 = exportRatchetPublicKey(peerIdentityPublic),
        sendChainKey = null,
        recvChainKey = null,
        sendMessageNumber = 0,
        recvMessageNumber = 0,
        previousSendChainLength = 0,
    )
}

/** Advance the DH ratchet: mix a fresh DH result into the root key, derive a new chain key. */
private fun dhRatchetStep(
    rootKey: ByteArray,
    myPrivate: X25519PrivateKeyParameters,
    theirPublic: X25519PublicKeyParameters,
): Pair<ByteArray, ByteArray> {
    val dh = ecdhBits(myPrivate, theirPublic)
    val out = hkdf(dh, rootKey, HKDF_INFO_DH, 64)
    return out.copyOfRange(0, 32) to out.copyOfRange(32, 64)
}

private fun pushSkipped(session: RatchetSession, entry: SkippedKey) {
    session.skippedKeys.add(entry)
    while (session.skippedKeys.size > MAX_SKIPPED_KEYS) session.skippedKeys.removeAt(0)
}

private fun takeSkipped(session: RatchetSession, ratchetPublicKeyB64: String, messageNumber: Int): ByteArray? {
    val idx = session.skippedKeys.indexOfFirst {
        it.ratchetPublicKeyB64 == ratchetPublicKeyB64 && it.messageNumber == messageNumber
    }
    if (idx == -1) return null
    val found = session.skippedKeys.removeAt(idx)
    return fromBase64(found.messageKeyB64)
}

/** Advance recvChainKey forward to `upTo`, caching each derived key as skipped (not returned). */
private fun skipForward(session: RatchetSession, ratchetPublicKeyB64: String, upTo: Int) {
    if (session.recvChainKey == null) return
    val count = upTo - session.recvMessageNumber
    if (count <= 0) return
    if (count > MAX_SKIP_PER_STEP) {
        throw RatchetError("refusing to derive $count skipped keys in one step")
    }
    for (i in 0 until count) {
        val (messageKey, nextChainKey) = stepChain(session.recvChainKey!!)
        pushSkipped(
            session,
            SkippedKey(
                ratchetPublicKeyB64 = ratchetPublicKeyB64,
                messageNumber = session.recvMessageNumber + i,
                messageKeyB64 = toBase64(messageKey),
            ),
        )
        session.recvChainKey = nextChainKey
    }
    session.recvMessageNumber = upTo
}

data class RatchetEncryptResult(val messageKey: SecretKey, val header: RatchetMessageHeader)

/** Sender side: derive the next message key, mutating the session in place. */
fun ratchetEncrypt(session: RatchetSession): RatchetEncryptResult {
    if (session.sendChainKey == null) {
        val theirPublic = importRatchetPublicKey(session.theirRatchetPublicKeyB64)
        val (newRootKey, chainKey) = dhRatchetStep(session.rootKey, session.myRatchetKeyPair.privateKey, theirPublic)
        session.rootKey = newRootKey
        session.sendChainKey = chainKey
        session.previousSendChainLength = session.sendMessageNumber
        session.sendMessageNumber = 0
    }
    val (messageKey, nextChainKey) = stepChain(session.sendChainKey!!)
    val header = RatchetMessageHeader(
        ratchetPublicKeyB64 = exportRatchetPublicKey(session.myRatchetKeyPair.publicKey),
        messageNumber = session.sendMessageNumber,
        previousChainLength = session.previousSendChainLength,
    )
    session.sendChainKey = nextChainKey
    session.sendMessageNumber += 1
    return RatchetEncryptResult(importAesKey(messageKey), header)
}

/**
 * Receiver side: derive the message key matching this header, mutating
 * the session in place (performing a DH ratchet step if the peer's key
 * changed — or if we've never received anything yet — and caching any
 * skipped-over keys for later out-of-order arrivals).
 */
fun ratchetDecrypt(session: RatchetSession, header: RatchetMessageHeader): SecretKey {
    val cached = takeSkipped(session, header.ratchetPublicKeyB64, header.messageNumber)
    if (cached != null) return importAesKey(cached)

    if (header.ratchetPublicKeyB64 != session.theirRatchetPublicKeyB64 || session.recvChainKey == null) {
        if (session.recvChainKey != null) {
            // Drain whatever's left of the OLD receiving chain (bounded by the
            // sender's declared previous-chain length) so late arrivals on it
            // can still be found in the skipped-key cache.
            skipForward(session, session.theirRatchetPublicKeyB64, header.previousChainLength)
        }
        val theirNewPublic = importRatchetPublicKey(header.ratchetPublicKeyB64)
        val (newRootKey, chainKey) = dhRatchetStep(session.rootKey, session.myRatchetKeyPair.privateKey, theirNewPublic)
        session.rootKey = newRootKey
        session.recvChainKey = chainKey
        session.recvMessageNumber = 0
        session.theirRatchetPublicKeyB64 = header.ratchetPublicKeyB64
        // Force a fresh sending step next time we send, with fresh ephemeral
        // material — this is what "heals" the session after a round trip.
        session.sendChainKey = null
        session.myRatchetKeyPair = generateRatchetKeyPair()
    }

    if (session.recvChainKey == null) {
        throw RatchetError("no receiving chain established for this header")
    }
    if (header.messageNumber < session.recvMessageNumber) {
        throw RatchetError(
            "message ${header.messageNumber} on the current chain was already consumed and is not in the skipped-key cache",
        )
    }
    if (header.messageNumber > session.recvMessageNumber) {
        skipForward(session, header.ratchetPublicKeyB64, header.messageNumber)
    }
    val (messageKey, nextChainKey) = stepChain(session.recvChainKey!!)
    session.recvChainKey = nextChainKey
    session.recvMessageNumber = header.messageNumber + 1
    return importAesKey(messageKey)
}
